### PR TITLE
Switch to oauthlib, add support for query string through command line API

### PR DESCRIPTION
**Pull request by diox**
This is to make the following work:

`% curling -i 'https://marketplace.allizom.org/api/v2/langpacks/?active=false'`

With the current code, curling actually ignores the query string and does a request to https://marketplace.allizom.org/api/v2/langpacks/, which is not what I need, so here is my attempt to fix it.

I also switched to oauthlib to get rid of the oauth2 dependency, which is insecure and unmaintained  : https://github.com/simplegeo/python-oauth2/issues/129 ; this could be done in a separate pull request if needed.

I know very little about curling, slumber, oauthlib vs oauth2, and although my patch seems to work for me, but it might have broken subtle things I don't understand/use yet, so take your time ;)

### CODE REVIEW

**Comment by andymckay** [comment]
Yeah this one might take a while, we use the same lib for communicating between solitude, webpay and zamboni, so we'd need to update those requirements. But oauth2 is the worst, so this will be worth it.

**Comment by diox** [comment]
Note: curling is the only reason zamboni requires oauth2, that was my main motivation to fix this.

File: curling/test.py [hunk 1]
```diff
@@ -128,8 +131,8 @@ def setUp(self):
         self.api = MockAPI('')
 
     def test_get_one(self):
-        eq_(self.api.services.settings('APPEND_SLASH').get_object(),
-            samples['GET:/services/settings/APPEND_SLASH/'])
+        eq_(self.api.services.setting.get_object(),
+            {'key': 'ABSOLUTE_URL_OVERRIDES'})
```
Comment (andymckay) [inline curling/test.py]: This changes the test a bit, the results `services.settings('APPEND_SLASH')` and `services.setting..` are different and this tests how a result without a meta in the result are handled. If that's covered elsewhere, that's cool.

File: curling/test.py [hunk 2]
```diff
@@ -149,10 +152,6 @@ def test_get_blank(self):
     def test_get_blank_fail(self):
         self.api.services.blankfail.get()
 
-    def test_get_empty(self):
-        res = self.api.services.nothing.get()
-        eq_(len(res), 0)
```
Comment (andymckay) [inline curling/test.py]: We don't like this test?

File: curling/test.py [hunk 3]
```diff
@@ -161,14 +160,6 @@ def test_get_empty(self):
     def test_get_many(self):
         self.api.services.settings.get_object()
 
-    def test_get_one(self):
-        eq_(self.api.services.setting.get_object(),
-            {'key': 'ABSOLUTE_URL_OVERRIDES'})
```
Comment (andymckay) [inline curling/test.py]: Ah `test_get_one` defined twice, now I see why.

File: curling/command.py [hunk 4]
```diff
@@ -80,9 +80,14 @@ def new(config, lib_api=None):
         return
 
     method = getattr(api, config.request.lower())
+    kwargs = {
+        'data': config.data
+    }
+    query_dict = OrderedDict(urlparse.parse_qsl(str(url.query)))
```
Comment (andymckay) [inline curling/command.py]: If `data` is in the query, this will override the other.

File: curling/command.py [hunk 5]
```diff
@@ -80,9 +80,14 @@ def new(config, lib_api=None):
...
+    kwargs.update(query_dict)
 
     try:
-        res = method(data=config.data)
+        res = method(**kwargs)
```
Comment (andymckay) [inline curling/command.py]: Should this be `method(data=config.data, params=query_dict)`?

File: curling/test.py (hunk 2 above)
Comment (diox) [inline curling/test.py]: Like `test_get_one`, it's actually defined twice, see below.

File: curling/test.py (hunk 3 above)
Comment (diox) [inline curling/test.py]: Yeah. I deleted the ones that weren't being run before.

File: curling/test.py (hunk 1 above)
Comment (diox) [inline curling/test.py]: Ah, maybe I misunderstood the test. Will double-check.

**Comment by andymckay** [comment]
The `method(data=config.data, params=query_dict)` is my only real concern. Other that r+wc.

I gave it a quick spin against some remote and local servers and it worked just fine. I wouldn't mind doing a full solitude test and the best way to do that is to release a package update all the things and then test.

Let's do a new release onto pypi and see! :)

File: curling/command.py (hunk 5 above)
Comment (diox) [inline curling/command.py]: So...

`method` points to the methods in the `API` class.

Those methods have this signature:
`def <name>(self, data, headers=None, **kwargs):`

When they make their request, they do this:
`
resp = self._request('<verb>', data=s.dumps(data),
                              headers=headers, params=kwargs)
`

So with the way I'm doing it, every key in `query_dict` will be passed through `kwargs` and end up in `params` in the end when making the actual slumber call and it works.

**However**, as you point out, that means `query_dict` can override `data` but also `headers`. This is bad, and needs to be fixed. I probably will call `method(config.data, headers=None, **query_dict)`. (And implement `-H` later to allow passing arbitrary headers...)

File: curling/test.py (hunk 1 above)
Comment (diox) [inline curling/test.py]: This `test_get_one` was not ran before, because the name clashes with another `test_get_one` test below. I tried changing the name and it didn't pass at all:
`AssertionError: {u'key': u'APPEND_SLASH'} != {'content': '{"key": "APPEND_SLASH"}'}`

Ideas?

File: curling/command.py (hunk 5 above)
Comment (diox) [inline curling/command.py]: So, I fixed it in the command, but because of the signature the methods in the lib (`def <name>(self, data, headers=None, **kwargs)`) this is still an issue. This is an issue that is already present in slumber, too - to fix it, the query dict would need to be passed as a separate parameter instead of accepting keyword arguments like this.

TL;DR: not a regression.

**Comment by diox** [comment]
Updated. See outdated diffs for additional comments.

**Comment by andymckay** [comment]
Awesome thanks, as we chatted on IRC, we'll tag and test solitude, zamboni and webpay out sometimes next week.
